java-checks/src/main/java/org/sonar/java/checks/ReleaseSensorsCheck.java

```diff
@@ -0,0 +1,115 @@
+ * <p>
+ * The implemented logic checks that after acquiring a sensor there is at least one release invocation within the same file.
+ */
+@Rule(key = "S6889")
+public class ReleaseSensorsCheck extends IssuableSubscriptionVisitor {
+
+  enum AcquireReleaseSensor {
+    LOCATION_MANAGER("android.location.LocationManager", "requestLocationUpdates", "removeUpdates"),
+    SENSOR_MANAGER("android.hardware.SensorManager", "registerListener", "unregisterListener"),
+    CAMERA("android.hardware.Camera", "open", RELEASE),
+    WIFI_MANAGER("android.net.wifi.WifiManager$MulticastLock", "acquire", RELEASE),
+    MEDIA_PLAYER("android.media.MediaPlayer", MethodMatchers.CONSTRUCTOR, RELEASE),
+    MEDIA_RECORDER("android.media.MediaRecorder", MethodMatchers.CONSTRUCTOR, RELEASE);
+
+    private final MethodMatchers acquireMethodMatcher;
+    private final MethodMatchers releaseMethodMatcher;
+
+    AcquireReleaseSensor(String sensorClass, String acquireMethod, String releaseMethod) {
+      this.acquireMethodMatcher = MethodMatchers.create().ofTypes(sensorClass).names(acquireMethod).withAnyParameters().build();
+      this.releaseMethodMatcher = MethodMatchers.create().ofTypes(sensorClass).names(releaseMethod).withAnyParameters().build();
+    }
+  }
+
+  private static class AcquireReleaseStatus {
+    List<Tree> acquireInvocations = new LinkedList<>();
+    boolean released = false;
+  }
+
+  public static final String RELEASE = "release";
+  private AcquireReleaseStatus[] statuses;
+
+  public ReleaseSensorsCheck() {
+    initStatuses();
+  }
+
+  private void initStatuses() {
+    this.statuses = IntStream.range(0, AcquireReleaseSensor.values().length)
+      .mapToObj(i -> new AcquireReleaseStatus())
+      .toArray(AcquireReleaseStatus[]::new);
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    // collect acquire invocations
+    Arrays.stream(AcquireReleaseSensor.values())
+      .filter(sensor -> isAcquireMethodInvocation(tree, sensor.acquireMethodMatcher))
+      .forEach(sensor -> statuses[sensor.ordinal()].acquireInvocations.add(tree));
+
+    // flag released invocations
+    Arrays.stream(AcquireReleaseSensor.values())
+      .filter(sensor -> isAcquireMethodInvocation(tree, sensor.releaseMethodMatcher))
+      .forEach(sensor -> statuses[sensor.ordinal()].released = true);
+  }
+
+  private static boolean isAcquireMethodInvocation(Tree tree, MethodMatchers methodMatchers) {
+    switch (tree.kind()) {
+      case METHOD_INVOCATION:
+        return methodMatchers.matches((MethodInvocationTree) tree);
+      case NEW_CLASS:
+        return methodMatchers.matches((NewClassTree) tree);
+      default:
+        return false;
+    }
+  }
+
+  @Override
+  public void leaveFile(JavaFileScannerContext context) {
```

Comment:
How about initializing in `setContext` instead and only clearing in `leaveFile`? This would avoid errors when rule execution runs into a problem mid-file, exiting with an exception and hence not executing `leaveFile`.

Reply:
It's a very good point! I'll keep it in mind for the future :)